Improved \not simplification

andreiburdusa opened the pull request:
Referencing Issue #1120 
---

###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`

andreiburdusa commented on kore/test/Test/Kore/Step/Simplification/Equals.hs:
```diff
@@ -249,9 +252,10 @@ test_equalsSimplification_Or_Pattern =
                                 , makeNotPredicate definedH
+                                , Simplification.Not.makePredicateNot
+                                    definedGWithSubstitution
                                 ]
```
I changed the order of elements in this list so the test would pass. I don't know if that was important

andreiburdusa commented:
Performance details from Jenkins:
{ "evm-semantics/tests/ethereum-tests/VMTests/vmArithmeticTest/add0.json": { "user_sec": 56.31, "resident_kbytes": 843872 } }
{ "evm-semantics/tests/ethereum-tests/VMTests/vmIOandFlowOperations/pop1.json": { "user_sec": 39.07, "resident_kbytes": 820356 } }
{ "evm-semantics/tests/interactive/sumTo10.evm": { "user_sec": 456.26, "resident_kbytes": 2861828 } }
{ "src/main/k/evm-semantics/sum-to-n/sum-to-n-spec.k": { "user_sec": 444.40, "resident_kbytes": 1086996 } }
{ "wasm-semantics/tests/proofs/simple-arithmetic-spec.k.prove": { "user_sec": 68.26, "resident_kbytes": 875508 } }
{ "wasm-semantics/tests/proofs/loops-spec.k.prove": { "user_sec": 94.91, "resident_kbytes": 689560 } }
{ "wasm-semantics/tests/proofs/memory-spec.k.prove": { "user_sec": 63.48, "resident_kbytes": 667556 }

It doesn't seem like the time increased too much

ttuegel commented on kore/test/Test/Kore/Step/Simplification/Not.hs:
```diff
@@ -46,11 +46,16 @@ test_simplifyEvaluated =
     , [substXA] `becomes_` [notEqualsXA]
     , [equalsXA, equalsXB] `becomes_` [neitherXAB]
     , [xAndEqualsXA] `becomes_` [termNotX, notEqualsXA]
+    , [termXAndY] `becomes_` [termNotXOrNotY]
```
The result here should be `[termNotX, termNotY]`.

ttuegel commented on kore/test/Test/Kore/Step/Simplification/Not.hs:
```diff
@@ -46,11 +46,16 @@ test_simplifyEvaluated =
     , [equalsXA, equalsXB] `becomes_` [neitherXAB]
     , [xAndEqualsXA] `becomes_` [termNotX, notEqualsXA]
+    , [termXAndY] `becomes_` [termNotXOrNotY]
+    , [termNotXAndY] `becomes_` [termXOrNotY]
```
The result here should be `[termX, termNotY]`.

ttuegel commented on kore/test/Test/Kore/Step/Simplification/Not.hs:
```diff
@@ -46,11 +46,16 @@ test_simplifyEvaluated =
+    , [termXAndY] `becomes_` [termNotXOrNotY]
+    , [termNotXAndY] `becomes_` [termXOrNotY]
+    , [notEqualsXA] `becomes_` [equalsXA]
+    , [notEqualsXAAndEqualsXB] `becomes_` [equalsXAOrNotEqualsXB]
```
```suggestion
    , [notEqualsXAAndEqualsXB] `becomes_` [equalsXA, notEqualsXB]
```

ttuegel commented on kore/test/Test/Kore/Step/Step.hs:
```diff
@@ -774,7 +777,7 @@ test_applyRewriteRulesParallel =
                         { term = initialTerm
                         , predicate =
-                            makeNotPredicate
+                            Simplification.Not.makePredicateNot
```
The test results should be written out here, instead of calling the simplifier.

virgil-serbanuta commented on kore/src/Kore/Step/Simplification/Not.hs:
```diff
@@ -174,11 +174,25 @@ makeTermNot
+makePredicateNot Syntax.Predicate.PredicateTrue =
+    Syntax.Predicate.makeFalsePredicate
+makePredicateNot genericPredicate =
+    fmap makeTermNot genericPredicate
```
See my comment above, this fmap always produces a `Syntax.Predicate`, although that shouldn't always be the case.

virgil-serbanuta commented on kore/src/Kore/Step/Simplification/Not.hs:
```diff
@@ -174,11 +174,25 @@ makeTermNot
 -- TODO: maybe other simplifications like
 -- not ceil = floor not
 -- not forall = exists not
+makeTermNot (Not_ _ term) = term
```
`term` might actually be an `Internal.Predicate` collapsed to a `Syntax.Predicate` in `makeEvaluatePredicate`. This function should return a simplified thing, that can't be simplified further, but simplifying the result again would produce the original `Internal.Predicate`

ttuegel commented on kore/src/Kore/Step/Simplification/Not.hs:
```diff
@@ -174,6 +179,9 @@ makeTermNot
 -- not forall = exists not
+makeTermNot (Not_ _ term) = term
+makeTermNot (And_ _ term1 term2) =
+    mkOr (makeTermNot term1) (makeTermNot term2)
```
Let's have this return a `MultiOr (TermLike variable)` instead.

andreiburdusa commented:
> This looks fine except for the comment I left, but I also want @virgil-serbanuta to re-review to see if double-negation simplification is handled the way we discussed.

I thought we agreed I will merge only what addresses your comments, because neither I nor @traiansf  know how to address @virgil-serbanuta 's

virgil-serbanuta commented on kore/src/Kore/Step/Simplification/Not.hs:
```diff
@@ -121,16 +123,16 @@ makeEvaluateNot
-    OrPattern.fromPatterns
-        [ Pattern.fromTermLike $ TermLike.markSimplified $ makeTermNot term
-        , Pattern.fromConditionSorted
+    merge
```
This looks a bit confusing at first sight, since `merge` is a fairly generic name. How about using it qualified (`OrPattern.merge`)? The same for `OrPattern.singleton`. I think that both are only used with qualifiers everywhere else.

andreiburdusa replied:
Done :+1:
